File: Templates/SolutionTemplate.cpp
/**
 * @file SolutionTemplate.cpp
 * @author Vaibhav Verma (https://github.com/Vaibhav1523)
 * @brief Production C++ Template for LeetCode / DSA Problem Solving
 * 
 * Target: Google / Amazon / Meta Technical Interview Standard
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <cassert>

class Solution {
public:
    /**
     * @brief Problem Solution Method
     * 
     * @param nums Input array
     * @param target Target value
     * @return std::vector<int> Result indices/values
     * 
     * @complexity
     * - Time Complexity:  O(N) -> Single pass algorithm
     * - Space Complexity: O(N) -> Hash map auxiliary storage
     */
    std::vector<int> solve(const std::vector<int>& nums, int target) {
        std::unordered_map<int, int> seen;
        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            int complement = target - nums[i];
            if (seen.find(complement) != seen.end()) {
                return {seen[complement], i};
            }
            seen[nums[i]] = i;
        }
        return {};
    }
};

// ================= AUTOMATED UNIT TESTS =================
void runTestSuite() {
    Solution sol;
    
    // Test Case 1: Standard Case
    std::vector<int> nums1 = {2, 7, 11, 15};
    int target1 = 9;
    std::vector<int> expected1 = {0, 1};
    assert(sol.solve(nums1, target1) == expected1);
    
    // Test Case 2: Duplicate Values
    std::vector<int> nums2 = {3, 3};
    int target2 = 6;
    std::vector<int> expected2 = {0, 1};
    assert(sol.solve(nums2, target2) == expected2);
    
    std::cout << "✅ [C++] All Unit Test Cases Passed!" << std::endl;
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    
    runTestSuite();
    return 0;
}

File: Templates/SolutionTemplate.java
/**
 * @file SolutionTemplate.java
 * @author Vaibhav Verma (https://github.com/Vaibhav1523)
 * @brief Production Java Template for LeetCode / DSA Problem Solving
 * 
 * Target: Google / Amazon / Meta Technical Interview Standard
 */

import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;

public class SolutionTemplate {

    /**
     * @brief Core Algorithm Implementation
     * 
     * @param nums Input array
     * @param target Target sum
     * @return int[] Indices of the two numbers adding to target
     * 
     * @complexity
     * - Time Complexity:  O(N) -> Linear sweep with HashMap lookup
     * - Space Complexity: O(N) -> Extra memory for frequency table
     */
    public int[] solve(int[] nums, int target) {
        if (nums == null || nums.length < 2) {
            return new int[]{};
        }

        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[]{map.get(complement), i};
            }
            map.put(nums[i], i);
        }
        return new int[]{};
    }

    // ================= AUTOMATED UNIT TESTS =================
    public static void main(String[] args) {
        SolutionTemplate sol = new SolutionTemplate();

        // Test Case 1
        int[] result1 = sol.solve(new int[]{2, 7, 11, 15}, 9);
        assert Arrays.equals(result1, new int[]{0, 1}) : "Test Case 1 Failed!";

        // Test Case 2
        int[] result2 = sol.solve(new int[]{3, 2, 4}, 6);
        assert Arrays.equals(result2, new int[]{1, 2}) : "Test Case 2 Failed!";

        System.out.println("✅ [Java] All Unit Test Cases Passed!");
    }
}
